### PR TITLE
Feature: add sc_scf_start to control drho before inner loop

### LINKED ISSUE #4067: Inner loop of DeltaSpin is confusing
### Describe the bug

[204_NO_KP_S2_deltaspin.zip](https://github.com/deepmodeling/abacus-develop/files/15141782/204_NO_KP_S2_deltaspin.zip)

```
Step (Outer -- Inner) =  4 -- 22          RMS = 4.082e-08     TIME(s) = 3.314e-01  
Meet convergence criterion ( < 1.000e-07 ), exit.       Total TIME(s) = 1.002e+01
after-optimization spin (uB): (print in the inner loop): 
ATOM 1         1.9999999911e+00
ATOM 2         2.0000000570e+00
after-optimization lambda (Ry/uB): (print in the inner loop): 
ATOM 1         7.8874261283e-02
ATOM 2         7.8874459986e-02
Inner optimization for lambda ends.
===============================================================================
 GE5    3.69e+00  4.02e+00  -6.819526e+03  7.532815e+00   8.351e-02  1.238e+01  
===============================================================================
Inner optimization for lambda begins ...
Covergence criterion for the iteration: 1.000e-07
initial lambda: 
ATOM 1         7.8874270322e-02
ATOM 2         7.8874458340e-02
initial spin: 
ATOM 1         1.9999999911e+00
ATOM 2         2.0000000570e+00
target spin: 
ATOM 1         2.0000000000e+00
ATOM 2         2.0000000000e+00
```

### Expected behavior

_No response_

### To Reproduce

_No response_

### Environment

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Verify the issue is not a duplicate.
- [ ] Describe the bug.
- [ ] Steps to reproduce.
- [ ] Expected behavior.
- [ ] Error message.
- [ ] Environment details.
- [ ] Additional context.
- [ ] Assign a priority level (low, medium, high, urgent).
- [ ] Assign the issue to a team member.
- [ ] Label the issue with relevant tags.
- [ ] Identify possible related issues.
- [ ] Create a unit test or automated test to reproduce the bug (if applicable).
- [ ] Fix the bug.
- [ ] Test the fix.
- [ ] Update documentation (if necessary).
- [ ] Close the issue and inform the reporter (if applicable).

### PR DESCRIPTION
**Pull request by hongriTianqi**
### Reminder
- [x] Have you linked an issue with this pull request?
- [x] Have you added adequate unit tests and/or case tests for your pull request?

### Linked Issue
Fix #4067 

### Unit Tests and/or Case Tests for my changes
- Updated related UT in input tests

### What's changed?
- The inner lambda loop will firstly calculate mag from rho_out to make the process more robust

### Any changes of core modules? (ignore if not applicable)
- The condition to invoke inner loop isis changed.
```
 // run the inner lambda loop to contrain atomic moments with the DeltaSpin method
    if (GlobalV::sc_mag_switch && iter > GlobalV::sc_scf_nmin && this->drho < GlobalV::sc_scf_start)
    {
        SpinConstrain<TK, psi::DEVICE_CPU>& sc = SpinConstrain<TK, psi::DEVICE_CPU>::getScInstance();
        sc.run_lambda_loop(iter-1);
    }
```

### CODE REVIEW

File: source/module_base/global_variable.cpp [hunk 1]
```diff
@@ -296,6 +296,7 @@ double sc_thr = 1.0e-6;
 int nsc = 100;
 int nsc_min = 2;
 int sc_scf_nmin = 2;
+double sc_scf_start = 0.1;
```
Comment (mohanchen) [inline source/module_base/global_variable.cpp:299]: do not add any GlobalV, use an alternative method.

File: source/module_esolver/esolver_ks_lcao.cpp [hunk 2]
```diff
@@ -689,9 +689,14 @@ void ESolver_KS_LCAO<TK, TR>::iter_init(const int istep, const int iter)
         this->p_hamilt->refresh();
     }
 
+    static bool drho_condition = false;
     // run the inner lambda loop to contrain atomic moments with the DeltaSpin method
-    if (GlobalV::sc_mag_switch && iter > GlobalV::sc_scf_nmin)
+    if (GlobalV::sc_mag_switch && iter > GlobalV::sc_scf_nmin && (this->drho < GlobalV::sc_scf_start || drho_condition))
```
Comment (mohanchen) [inline source/module_esolver/esolver_ks_lcao.cpp:694]: GlobalV will be removed in near future.

File: source/module_base/global_variable.cpp (hunk 1 above)
Comment (hongriTianqi) [reply source/module_base/global_variable.cpp:299]: @mohanchen Thanks for your comments. I will make a new PR to achieve this. So close this one now.

File: source/module_esolver/esolver_ks_lcao.cpp (hunk 2 above)
Comment (hongriTianqi) [reply source/module_esolver/esolver_ks_lcao.cpp:694]: @mohanchen Thanks for your comments. I will make a new PR to achieve this. So close this one now.

**Pull request closed without merging**